dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ClientRateLimitHelper.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * <ul>
+ *     <li>RateLimit-Limit</li>
+ *     <li>RateLimit-Remaining</li>
+ *     <li>RateLimit-Reset</li>
+ * </ul>
+ */
+public class ClientRateLimitHelper {
+    // public static final String LIMIT_HEADER = "X-RateLimit-Limit";
+    public static final String REMAINING_HEADER = "X-RateLimit-Remaining";
+    public static final String RESET_HEADER = "X-RateLimit-Reset";
+    /**
+     * The padding to keep from 0 remaining requests. If the &quot;X-RateLimit-Remaining&quut; header has a value
+     * smaller than this, wait for a rate limit reset before proceeding. This
+     */
+    private static final int REQUESTS_REMAINING_PADDING = 10;
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClientRateLimitHelper.class);
+    private final Duration maxWait;
+
+    /**
+     * Creates a helper, specifying how long to wait for a ratelimit reset.
+     * @param maxWait how long to wait for a reset
+     */
+    public ClientRateLimitHelper(Duration maxWait) {
+        this.maxWait = maxWait;
+    }
+
+    /**
+     * Checks if rate limit headers are present. If they are present, and the remaining limit is less than {@code BACK_OFF}, the method
+     * sleeps until the reset time is hit. Otherwise, does nothing.
+     *
+     * <p>If the wait for the reset time would be more than the <code>maxWait</code> parameter passed to the constructor, then it throws
+     * a <code>CustomWebApplicationException</code>.</p>
+     *
+     * @param headers the headers of an HTTP response
+     */
+    public void checkRateLimit(Map<String, List<String>> headers) {
+        getRemainingAndReset(headers).ifPresent(remainingAndReset -> {
+            if (remainingAndReset.remaining < REQUESTS_REMAINING_PADDING) {
+                LOG.info(headers.toString());
+                final Instant now = Instant.now();
+                LOG.info("Now is {}", now);
+                final Duration durationToReset = Duration.between(now, remainingAndReset.resetTime);
+                LOG.info("Duration is {}", durationToReset);
+                if (durationToReset.compareTo(maxWait) > 0) {
+                    throw new CustomWebApplicationException("Rate limits on an external service hit", HttpStatus.SC_INTERNAL_SERVER_ERROR);
+                }
+                if (!durationToReset.isNegative()) {
+                    try {
+                        LOG.info("Sleeping for {} seconds", durationToReset.toSeconds());
+                        Thread.sleep(durationToReset.toMillis());
+                    } catch (InterruptedException e) {
+                        LOG.error("Interrupted exception while waiting for reset limit", e);
+                    }
+                }
+            }
+        });
+    }
+
+    private Optional<RemainingAndReset> getRemainingAndReset(Map<String, List<String>> headers) {
+        final Map<String, List<String>> map = new CaseInsensitiveMap(headers);
```

Comment:
Do we need `CaseInsensitiveMap<>` here?

Reply:
To be safe, I think so. HTTP headers are [case-insensitive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers). Although the Zenodo headers are currently coming back all lower-case, I don't think it's guaranteed that they have to be.

Reply:
> To be safe, I think so. HTTP headers are [case-insensitive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers). Although the Zenodo headers are currently coming back all lower-case, I don't think it's guaranteed that they have to be.

Oh, yes, functionally, `CaseInsensitiveMap` does the right thing.  I was referring to `CaseInsensitiveMap` versus `CaseInsensitiveMap<>`. the first is a raw type and the second is a parameterized type: https://www.baeldung.com/raw-types-java

Reply:
🤦